### PR TITLE
Feature: RWA-532 && RWA-310: Implement support for multiple categories witch Backwards compatibility

### PR DESCRIPTION
**Pull request by alexbaeza**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/RWA-532
https://tools.hmcts.net/jira/browse/RWA-310

### Change description ###

Implement support for multiple categories when initiating a task
Implement support for multiple categories when cancelling a task
Implement support for multiple categories when warning a task
Migrate to feign client.
Remove use of rest template
Remove unnecessary rest template clients
Remove CancelTaskException and Advice as it was not being used
Use a more generic approach to handle dmn values
Use more generic for responses using marker interfaces
Minor class package refactorings
Clean and updated unit test
Clean up Integration test

**Does this PR introduce a breaking change?** (check one with "x")

```
[ x] Yes
[ ] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/wacaseeventhandler/domain/camunda/response/CancellationEvaluateResponse.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import lombok.Builder;
+import lombok.ToString;
+import uk.gov.hmcts.reform.wacaseeventhandler.domain.camunda.DmnValue;
+
+@ToString
+@Builder
+@SuppressWarnings("PMD.UseConcurrentHashMap")
```
Comment (iguannago) [inline src/main/java/uk/gov/hmcts/reform/wacaseeventhandler/domain/camunda/response/CancellationEvaluateResponse.java:11]: a small one, but do you need to suppress this here, it does not seem you use any Map at all

File: src/main/java/uk/gov/hmcts/reform/wacaseeventhandler/domain/camunda/response/InitiateEvaluateResponse.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import lombok.Builder;
+import lombok.ToString;
+import uk.gov.hmcts.reform.wacaseeventhandler.domain.camunda.DmnValue;
+
+@ToString
+@Builder
+@SuppressWarnings("PMD.UseConcurrentHashMap")
```
Comment (iguannago) [inline src/main/java/uk/gov/hmcts/reform/wacaseeventhandler/domain/camunda/response/InitiateEvaluateResponse.java:12]: same here

File: src/main/java/uk/gov/hmcts/reform/wacaseeventhandler/domain/ccd/message/EventInformation.java [hunk 3]
```diff
@@ -7,15 +7,13 @@
...
 import lombok.Builder;
-import lombok.EqualsAndHashCode;
 import lombok.ToString;
 
 import java.time.LocalDateTime;
 import java.util.Locale;
 import javax.validation.constraints.NotEmpty;
 
 @ToString
-@EqualsAndHashCode
```
Comment (iguannago) [inline src/main/java/uk/gov/hmcts/reform/wacaseeventhandler/domain/ccd/message/EventInformation.java:18]: Why are we removing this annotation? This is usually a good thing in the Pojo to ensure uniqueness

File: src/main/java/uk/gov/hmcts/reform/wacaseeventhandler/handlers/CancellationCaseEventHandler.java [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+                );
+            });
+    }
+
+    private EvaluateDmnRequest buildEvaluateDmnRequest(
+        String previousStateId,
+        String eventId,
+        String newStateId
+    ) {
+        Map<String, DmnValue<?>> variables = Map.of(
```
Comment (iguannago) [inline src/main/java/uk/gov/hmcts/reform/wacaseeventhandler/handlers/CancellationCaseEventHandler.java:89]: Small suggestion: I think is more thread-safe and more performant to use the concurrent Hash map when possible. And you can get rid of that suppression at the top-level class too. 
I like to write  Map.of , though.

File: src/main/java/uk/gov/hmcts/reform/wacaseeventhandler/handlers/CancellationCaseEventHandler.java [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+     * @return The message request object.
+     */
+    private SendMessageRequest createCancellationMessage(String caseReference,
+                                                         DmnValue<String> categories,
+                                                         DmnValue<String> processCategories) {
+
+        Map<String, DmnValue<?>> correlationKeys = new HashMap<>();
+        correlationKeys.put("caseId", dmnStringValue(caseReference));
+
+        if (categories != null && categories.getValue() != null) {
```
Comment (iguannago) [inline src/main/java/uk/gov/hmcts/reform/wacaseeventhandler/handlers/CancellationCaseEventHandler.java:142]: suggestion: This block and the block below has the same implementation. If you select this block and refactor it into a method you can then re-use that method for the block below.

**Review by iguannago — COMMENTED** [review]
Few minor suggestions. Happy to approve if you want, let me know.

**Comment by iguannago** [comment]
A few minor suggestions, Alex. Happy to approve it if you want to.

**Review by alexbaeza — COMMENTED** [review]
applied suggestions

**Review by iguannago — DISMISSED** [review]

File: src/functionalTest/java/uk/gov/hmcts/reform/wacaseeventhandler/controllers/CaseEventHandlerControllerTest.java [hunk 6]
```diff
@@ -565,13 +565,20 @@ public void given_initiation_task_with_followup_overdue_ctg_when_cancelled_with_
     public void given_event_requestHearingRequirementsFeature_when_initiated_verfiy_task_creation() {
         String caseId1 = UUID.randomUUID().toString();
         final String taskId = initiateTaskForGivenId(caseId1, "requestHearingRequirementsFeature",
```
Comment (BenNzewi) [inline src/functionalTest/java/uk/gov/hmcts/reform/wacaseeventhandler/controllers/CaseEventHandlerControllerTest.java:567]: This is calling initiateTaskForGivenId with the wrong number of args. (The method has 5 args in the implementation below). Causes compilation error on CEH

Comment (alexbaeza) [reply src/functionalTest/java/uk/gov/hmcts/reform/wacaseeventhandler/controllers/CaseEventHandlerControllerTest.java:567]: Fixed sorry something must have gone wrong on the rebase

**Review by iguannago — APPROVED** [review]

**Pull request merged**